src/ucs/datastruct/ptr_map.inl

```diff
@@ -119,28 +208,69 @@ ucs_ptr_map_put(ucs_ptr_map_t *map, void *ptr, int indirect,
  *
  * @return UCS_OK if found, UCS_ERR_NO_PROGRESS if this key is direct and
  *         therefore no action was performed, UCS_ERR_NO_ELEM if not found.
+ * 
+ * @note The method moves pointer to unsafe hash table, if the pointer was found
+ * in the thread safe table.
  */
 static UCS_F_ALWAYS_INLINE ucs_status_t
 ucs_ptr_map_get(ucs_ptr_map_t *map, ucs_ptr_map_key_t key, int extract,
                 void **ptr_p)
 {
-    khiter_t iter;
+    ucs_status_t status;
 
-    if (ucs_likely(!ucs_ptr_map_key_indirect(key))) {
-        *ptr_p = (void*)key;
+    if (ucs_likely(ucs_ptr_check_direct_get(key, ptr_p))) {
         return UCS_ERR_NO_PROGRESS;
     }
 
-    iter = kh_get(ucs_ptr_map_impl, &map->hash, key);
-    if (ucs_unlikely(iter == kh_end(&map->hash))) {
-        return UCS_ERR_NO_ELEM;
+    status = ucs_ptr_hash_get(&map->hash, key, extract, ptr_p);
+    if (ucs_likely(status != UCS_ERR_NO_ELEM)) {
+        return status;
     }
 
-    *ptr_p = kh_value(&map->hash, iter);
-    if (extract) {
-        kh_del(ucs_ptr_map_impl, &map->hash, iter);
+    ucs_spin_lock(&map->lock);
+    status = ucs_ptr_hash_get(&map->hash_safe, key, 1, ptr_p);
+    ucs_spin_unlock(&map->lock);
+
+    if ((status != UCS_OK) || extract) {
+        return status;
     }
-    return UCS_OK;
+
+    return ucs_ptr_hash_put(&map->hash, key, *ptr_p);
```

Comment:
`ucs_ptr_hash_put` may return `ALREADY_EXISTS`, which would be weird to get from the the `get` operation. I'd rearrange the code somehow to make sure that only proper error codes are returned

Reply:
We can reach `ucs_ptr_hash_put` only if `ucs_ptr_hash_get` from line 225 returns `UCS_ERR_NO_ELEM`.
This means that `ucs_ptr_hash_put` cannot return `UCS_ERR_ALREADY_EXISTS`.

Reply:
can it be added by a different thread after `ucs_ptr_hash_get` returns `no_elem`? 
I'd at least add assert here

Reply:
We will face different issue, if `ucs_ptr_hash_get(&map->hash, key, extract, ptr_p)` in a different thread returns `UCS_ERR_NO_ELEM`.
We can get `UCS_ERR_NO_ELEM` from `ucs_ptr_hash_get(&map->hash_safe, key, 1, ptr_p)`. Because the key now is in  `map->hash`.

Reply:
why getter API whic assumed to be "read only" does "put" in general?

Reply:
The idea is to use auxiliary cache to have lock-free get operation except the first get. Also mentioned here https://github.com/openucx/ucx/pull/7159#discussion_r680802048